# Stack Overflow: Tag Prediction

**1. Business Problem**

**1.1 Description**

Description

Stack Overflow is the largest, most trusted online community for developers to learn, share their programming knowledge, and build their careers.

Stack Overflow is something which every programmer use one way or another. Each month, over 50 million developers come to Stack Overflow to learn, share their knowledge, and build their careers. It features questions and answers on a wide range of topics in computer programming. The website serves as a platform for users to ask and answer questions, and, through membership and active participation, to vote questions and answers up or down and edit questions and answers in a fashion similar to a wiki or Digg. As of April 2014 Stack Overflow has over 4,000,000 registered users, and it exceeded 10,000,000 questions in late August 2015. Based on the type of tags assigned to questions, the top eight most discussed topics on the site are: Java, JavaScript, C#, PHP, Android, jQuery, Python and HTML.

**Problem Statemtent**

Suggest the tags based on the content that was there in the question posted on Stackoverflow.

Source: https://www.kaggle.com/c/facebook-recruiting-iii-keyword-extraction/

**1.2 Source / useful links**

Data Source : https://www.kaggle.com/c/facebook-recruiting-iii-keyword-extraction/data
Youtube : https://youtu.be/nNDqbUhtIRg
Research paper : https://www.microsoft.com/en-us/research/wp-content/uploads/2016/02/tagging-1.pdf
Research paper : https://dl.acm.org/citation.cfm?id=2660970&dl=ACM&coll=DL

**1.3 Real World / Business Objectives and Constraints**

1. Predict as many tags as possible with high precision and recall. --> If I missed a tag for a data point (Title+Description) my recall decreases. And if I provide incorrect tag to a data point my Precision decreases. 
2. Incorrect tags could impact customer experience on StackOverflow. 
3. No strict latency constraints.

**2. Machine Learning problem**

**2.1 Data**

2.1.1 Data Overview

All of the data is in 2 files: Train and Test.

Train.csv contains 4 columns: Id, Title, Body, Tags.

--> Id - Unique identifier for each question
--> Title - The question's title
--> Body - The body of the question
--> Tags - The tags associated with the question (all lowercase, should not contain tabs '\t' or ampersands '&')

Test.csv contains the same columns but without the Tags, which you are to predict.
Size of Train.csv - 6.75GB
Size of Test.csv - 2GB
Number of rows in Train.csv = 6034195

2.1.2 Example Data point

Title:  Implementing Boundary Value Analysis of Software Testing in a C++ program?
Body:
        #include&lt;
        iostream&gt;\n
        #include&lt;
        stdlib.h&gt;\n\n
        using namespace std;\n\n
        int main()\n
        {\n
                 int n,a[n],x,c,u[n],m[n],e[n][4];\n         
                 cout&lt;&lt;"Enter the number of variables";\n         cin&gt;&gt;n;\n\n         
                 cout&lt;&lt;"Enter the Lower, and Upper Limits of the variables";\n         
                 for(int y=1; y&lt;n+1; y++)\n         
                 {\n                 
                    cin&gt;&gt;m[y];\n                 
                    cin&gt;&gt;u[y];\n         
                 }\n         
                 for(x=1; x&lt;n+1; x++)\n         
                 {\n                 
                    a[x] = (m[x] + u[x])/2;\n         
                 }\n         
                 c=(n*4)-4;\n         
                 for(int a1=1; a1&lt;n+1; a1++)\n         
                 {\n\n             
                    e[a1][0] = m[a1];\n             
                    e[a1][1] = m[a1]+1;\n             
                    e[a1][2] = u[a1]-1;\n             
                    e[a1][3] = u[a1];\n         
                 }\n         
                 for(int i=1; i&lt;n+1; i++)\n         
                 {\n            
                    for(int l=1; l&lt;=i; l++)\n            
                    {\n                 
                        if(l!=1)\n                 
                        {\n                    
                            cout&lt;&lt;a[l]&lt;&lt;"\\t";\n                 
                        }\n            
                    }\n            
                    for(int j=0; j&lt;4; j++)\n            
                    {\n                
                        cout&lt;&lt;e[i][j];\n                
                        for(int k=0; k&lt;n-(i+1); k++)\n                
                        {\n                    
                            cout&lt;&lt;a[k]&lt;&lt;"\\t";\n               
                        }\n                
                        cout&lt;&lt;"\\n";\n            
                    }\n        
                 }    \n\n        
                 system("PAUSE");\n        
                 return 0;    \n
        }\n
        </code></pre>\n\n
        <p>The answer should come in the form of a table like</p>\n\n
        <pre><code>       
        1            50              50\n       
        2            50              50\n       
        99           50              50\n       
        100          50              50\n       
        50           1               50\n       
        50           2               50\n       
        50           99              50\n       
        50           100             50\n       
        50           50              1\n       
        50           50              2\n       
        50           50              99\n       
        50           50              100\n
        </code></pre>\n\n
        <p>if the no of inputs is 3 and their ranges are\n
        1,100\n
        1,100\n
        1,100\n
        (could be varied too)</p>\n\n
        <p>The output is not coming,can anyone correct the code or tell me what\'s wrong?</p>\n'
Tags: 'c++ c'

Points to be noted after seeing the data point:
--> The tags are separated by the space "c++ c"
--> The title is less verbose than body. 
--> The body can contain a lot of code.

**2.2 Mapping the real-world problem to a Machine Learning Problem**

**2.2.1 Type of Machine Learning Problem**

It is a **multi-label classification** problem. Multi-label Classification: Multilabel classification assigns to each sample a set of target labels. This can be thought as predicting properties of a data-point that are not mutually exclusive, such as topics that are relevant for a document. A question on Stackoverflow might be about any of C, Pointers, FileIO and/or memory-management at the same time or none of these.
__Credit__: http://scikit-learn.org/stable/modules/multiclass.html

**2.2.2 Performance metric**

Before going into the details we must know the metric system we are going to use to check the performance of our model. 
+ We can't use the conventional f1 score. Because it is for binary-class classification and not for multi-class and multi-label classification problem.  
+ What we can do is to use one vs rest approach to calculate the f1 score for each individual classes. 

**Micro-Averaged F1-Score (Mean F Score)** : The F1 score can be interpreted as a weighted average of the precision and recall, where an F1 score reaches its best value at 1 and worst score at 0. The relative contribution of precision and recall to the F1 score are equal. The formula for the F1 score is:

F1 = 2 * (precision * recall) / (precision + recall)

In the multi-class and multi-label case, this is the weighted average of the F1 score of each class.

**'Micro f1 score':**
Calculate metrics globally by counting the total true positives, false negatives and false positives. This is a better metric when we have class imbalance.  --> It consider the number of occurrence of the classes in the data (weighted f1). 

**'Macro f1 score':**
Calculate metrics for each label, and find their unweighted mean. This does not take label imbalance into account.

https://www.kaggle.com/wiki/MeanFScore
http://scikit-learn.org/stable/modules/generated/sklearn.metrics.f1_score.html

**Hamming loss** : The Hamming loss is the fraction of labels that are incorrectly predicted.
https://www.kaggle.com/wiki/HammingLoss

![f1 score](images/image1.png)

Hamming loss : The Hamming loss is the fraction of labels that are incorrectly predicted.
https://www.kaggle.com/wiki/HammingLoss

![Hamming Loss](images/image2.png)

**3. Exploratory Data Analysis**

**3.1 Data Loading and Cleaning**

3.1.1 Using Pandas with SQLite to Load the data

In [2]:
from datetime import datetime
import os
import sqlite3
from sqlalchemy import create_engine
import pandas as pd

if not os.path.isfile('data/dataset/train.db'):
    start = datetime.now()
    disl_engine = create_engine('sqlite:///data/dataset/train.db')
    start = datetime.now()
    chunksize = 180000
    j = 0
    index_start = 1
    for df in pd.read_csv('data/dataset/Train.csv', names=['Id', 'Title', 'Body', 'Tags'], chunksize=chunksize, iterator=True, encoding='utf-8'):
        df.index += index_start
        j += 1
        print(f'{j*chunksize} rows processed')
        df.to_sql('data', disl_engine, if_exists='append')
        index_start = df.index[-1] + 1
    print(f'Time taken to run this cell: {datetime.now() - start}')

/var/folders/9c/7jq79dg16qqdqst791mgh63w0000gn/T/ipykernel_34121/910857947.py:14: DtypeWarning: Columns (0) have mixed types. Specify dtype option on import or set low_memory=False.
  for df in pd.read_csv('data/dataset/Train.csv', names=['Id', 'Title', 'Body', 'Tags'], chunksize=chunksize, iterator=True, encoding='utf-8'):


180000 rows processed
360000 rows processed
540000 rows processed
720000 rows processed
900000 rows processed
1080000 rows processed
1260000 rows processed
1440000 rows processed
1620000 rows processed
1800000 rows processed
1980000 rows processed
2160000 rows processed
2340000 rows processed
2520000 rows processed
2700000 rows processed
2880000 rows processed
3060000 rows processed
3240000 rows processed
3420000 rows processed
3600000 rows processed
3780000 rows processed
3960000 rows processed
4140000 rows processed
4320000 rows processed
4500000 rows processed
4680000 rows processed
4860000 rows processed
5040000 rows processed
5220000 rows processed
5400000 rows processed
5580000 rows processed
5760000 rows processed
5940000 rows processed
6120000 rows processed
Time taken to run this cell: 0:01:52.483691


3.1.2 Counting the number of rows

In [4]:
if os.path.isfile('data/dataset/train.db'):
    start = datetime.now()
    con = sqlite3.connect('data/dataset/train.db')
    num_rows = pd.read_sql_query("SELECT count(*) FROM data", con)
    #Always remember to close the database
    print(f'Number of rows in the database: {num_rows["count(*)"].values[0]}')
    con.close()
    print(f'Time taken to run this cell: {datetime.now() - start}')
else:
    print(f'Please download the data')

Number of rows in the database: 6034196
Time taken to run this cell: 0:00:02.946916


3.1.3 Checking for duplicates

In [5]:
if os.path.isfile('data/dataset/train.db'):
    start = datetime.now()
    con = sqlite3.connect('data/dataset/train.db')
    df_no_dup = pd.read_sql_query("SELECT Title, Body, Tags, COUNT(*) as cnt_dup FROM data GROUP BY Title, Body, Tags", con)
    con.close()
    print(f'Time taken to run this cell: {datetime.now() - start}')
else:
    print(f'Please download the data')

Time taken to run this cell: 0:07:28.432139


In [6]:
df_no_dup.head()
# The duplicates are present in the data

Title  \
0       Implementing Boundary Value Analysis of S...   
1           Dynamic Datagrid Binding in Silverlight?   
2           Dynamic Datagrid Binding in Silverlight?   
3      java.lang.NoClassDefFoundError: javax/serv...   
4      java.sql.SQLException:[Microsoft][ODBC Dri...   

                                                Body  \
0  <pre><code>#include&lt;iostream&gt;\n#include&...   
1  <p>I should do binding for datagrid dynamicall...   
2  <p>I should do binding for datagrid dynamicall...   
3  <p>I followed the guide in <a href="http://sta...   
4  <p>I use the following code</p>\n\n<pre><code>...   

                                  Tags  cnt_dup  
0                                c++ c        1  
1          c# silverlight data-binding        1  
2  c# silverlight data-binding columns        1  
3                             jsp jstl        1  
4                            java jdbc        2

In [10]:
print(f'Number of duplicate rows: {num_rows["count(*)"].values[0] - df_no_dup.shape[0]} and the percentage: {(1 - df_no_dup.shape[0]/num_rows["count(*)"].values[0])*100}%')

Number of duplicate rows: 1827881 and the percentage: 30.292038906260256%


If we want to know how many times a question repeated in the dataset

In [11]:
df_no_dup.cnt_dup.value_counts()

cnt_dup
1    2656284
2    1272336
3     277575
4         90
5         25
6          5
Name: count, dtype: int64

So, here we can see number of question repeated one time is 2656284 and so on. 